### PR TITLE
Add kickstart tar installation test

### PR DESCRIPTION
**Pull request by jikortus**
Marked as [WIP] as it hasn't been fully tested yet due to [RHBZ1724181](https://bugzilla.redhat.com/show_bug.cgi?id=1724181).

### CODE REVIEW

File: tests/cli/test_install_tar.sh [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+name = "grubby-deprecated"
+version = "*"
+
+[[packages]]
+name = "kernel"
+version = "*"
+
+[[packages]]
+name = "openssh-server"
+version = "7.*"
```
Comment (atodorov) [inline tests/cli/test_install_tar.sh]: `*` is fine for a version. This already differs in downstream branches I think.

File: tests/cli/test_install_tar.sh [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+[[groups]]
+name = "anaconda-tools"
+
+[[customizations.user]]
+name = "root"
+key = "$pub_key"
+__EOF__
+        composer-cli blueprints push $blueprint
```
Comment (atodorov) [inline tests/cli/test_install_tar.sh]: Use "$CLI"

File: tests/cli/test_install_tar.sh [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+        image="$uuid-root.tar.xz"
+    rlPhaseEnd
+
+    rlPhaseStartTest "Install tar image using kickstart liveimg command"
+        cat >> $ks_path << __EOF__
+%post
+echo -e "curl http://10.0.2.2/image-$uuid-booted\n" >> /etc/rc.d/rc.local
+chmod +x /etc/rc.d/rc.local
+%end
+__EOF__
```
Comment (atodorov) [inline tests/cli/test_install_tar.sh]: You can create the entire kickstart file here, instead of splitting it in 2 parts.

File: tests/cli/test_install_tar.sh [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+        mv $image $image_path
+        restorecon $image_path
+        $QEMU_BIN -m 2048 -drive file=disk.img,format=raw -nographic -monitor none -net user,id=nic0,hostfwd=tcp::2222-:22 \
+            -net nic -kernel vmlinuz -initrd initrd.img \
+            -append "inst.ks=http://10.0.2.2/ks-tar.cfg inst.stage2=$baseurl console=ttyS0" &
+        timeout=$((60*60)) # wait up to 60 minutes
+        t=0
+        while ((t < timeout)) && ! grep -q "image-$uuid-booted" /var/log/httpd/access_log; do
+            rlLogInfo "Waiting for the tar image to be installed and system to boot up."
+            sleep 10
```
Comment (atodorov) [inline tests/cli/test_install_tar.sh]: IMO you can increase sleeps to 60 everywhere b/c these operations take a long time, though not strictly necessary.

File: tests/cli/test_install_tar.sh [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+lang en_US.UTF-8
+timezone America/New_York
+keyboard us
+rootpw redhat
+bootloader --location=mbr
+zerombr
+clearpart --initlabel --all
+autopart
+
+liveimg --url http://10.0.2.2/root.tar.xz
```
Comment (atodorov) [inline tests/cli/test_install_tar.sh]: Is this IP address a stable interface ? With libvirtd on RHEL the default is 192.168.122.1, never seen this one. Maybe you can also grab the actual address of the system and use it (if it is not overly complicated). Whichever one is good as long as we know it is stable.

File: tests/cli/test_install_tar.sh [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+key = "$pub_key"
+__EOF__
+        composer-cli blueprints push $blueprint
+        image_path="/var/www/html/root.tar.xz"
+
+        version=$(awk -F = '$1 == "VERSION_ID" { print $2 }' /etc/os-release)
+        arch=$(uname -m)
+        baseurl=$(curl "https://mirrors.fedoraproject.org/mirrorlist?repo=fedora-${version}&arch=${arch}" | \
+            grep -v "^#" | head -n 1)
+        curl --remote-name-all $baseurl/images/pxeboot/{vmlinuz,initrd.img}
```
Comment (atodorov) [inline tests/cli/test_install_tar.sh]: @bcl I am wondering if lorax-composer keeps a reference (or has an easy access to that reference) to the kernel/initrd used during building the image (i.e. from anaconda)?

If we could grab these from a local directory referenced by the compose UUID it would be much more easier instead of trying to figure out what distro we are running on and where to get these files from.

File: tests/cli/test_install_tar.sh (hunk 5 above)
Comment (larskarlitski) [reply tests/cli/test_install_tar.sh]: `file:///` urls work as well.

File: tests/cli/test_install_tar.sh [hunk 7]
```diff
@@ -0,0 +1,142 @@
+#!/bin/bash
+# Note: execute this file from the project root directory
+
+#####
+#
+# Builds tar image and install it using liveimg kickstart command
```
Comment (larskarlitski) [inline tests/cli/test_install_tar.sh]: What exactly is this testing? Is this something people do?

Comment (atodorov) [reply tests/cli/test_install_tar.sh]: We don't have accurate information to answer this question but this scenario was identified when we did the original assessment for Composer and has always been part of our integration test suite. So far we've executed it manually.

Comment (atodorov) [reply tests/cli/test_install_tar.sh]: FTR my understanding is that using tarball images to deploy a system is somewhat common, Brian or Terry may know more.

File: tests/cli/test_install_tar.sh (hunk 5 above)
Comment (atodorov) [reply tests/cli/test_install_tar.sh]: What do you mean by that ? If I understand you correctly the tarball image will need to be made available via partition on the hard-drive before we do the installation. Not really easier than firing up httpd and getting the IP address right.

File: tests/cli/test_install_tar.sh (hunk 7 above)
Comment (larskarlitski) [reply tests/cli/test_install_tar.sh]: Thanks, that helps. I don't see the use-case myself, but I'll investigate.

File: tests/cli/test_install_tar.sh (hunk 5 above)
Comment (larskarlitski) [reply tests/cli/test_install_tar.sh]: Heh, sorry for the brevity. I mistook the for a kickstart we're running from the host, in which case absolute `file://` urls would have worked. This one is passed to a booted system though, so your approach is totally valid. Thanks!

File: tests/cli/test_install_tar.sh (hunk 6 above)
Comment (bcl) [reply tests/cli/test_install_tar.sh]: If the blueprint includes the kernel the final-kickstart.ks will have the version in it.

**Comment by bcl** [comment]
I saw a comment about IP addresses, and can't tell if it was answered. You can set the subnet to use by passing in `net=192.168.103.0/24` as one of the netdev options.

File: tests/cli/test_compose_tar_kickstart.sh [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+        $CLI blueprints push $blueprint
+        image_path="/var/www/html/root.tar.xz"
+
+        version=$(awk -F = '$1 == "VERSION_ID" { print $2 }' /etc/os-release | tr -d \")
+        arch=$(uname -m)
+        baseurl=$(curl "https://mirrors.fedoraproject.org/mirrorlist?repo=fedora-${version}&arch=${arch}" | \
+            grep -v "^#" | head -n 1)
+        curl --remote-name-all $baseurl/images/pxeboot/{vmlinuz,initrd.img}
+
+        fallocate -l 5G disk.img
```
Comment (atodorov) [inline tests/cli/test_compose_tar_kickstart.sh]: surround with `rlRun` so we get a failure status if the system doesn't have enough free disk space

File: tests/cli/test_compose_tar_kickstart.sh [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+
+__EOF__
+        $CLI blueprints push $blueprint
+        image_path="/var/www/html/root.tar.xz"
+
+        version=$(awk -F = '$1 == "VERSION_ID" { print $2 }' /etc/os-release | tr -d \")
+        arch=$(uname -m)
+        baseurl=$(curl "https://mirrors.fedoraproject.org/mirrorlist?repo=fedora-${version}&arch=${arch}" | \
+            grep -v "^#" | head -n 1)
+        curl --remote-name-all $baseurl/images/pxeboot/{vmlinuz,initrd.img}
```
Comment (atodorov) [inline tests/cli/test_compose_tar_kickstart.sh]: IMO use rlRun so we can get errors reported in case of networking issues

File: tests/cli/test_compose_tar_kickstart.sh [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+        uuid=$($CLI compose start $bp_name tar)
+        rlAssertEquals "exit code should be zero" $? 0
+
+        uuid=$(echo $uuid | cut -f 2 -d' ')
+    rlPhaseEnd
+
+    rlPhaseStartTest "compose finished"
+        if [ -n "$uuid" ]; then
+            until $CLI compose info $uuid | grep 'FINISHED\|FAILED'; do
+                sleep 10
```
Comment (atodorov) [inline tests/cli/test_compose_tar_kickstart.sh]: You can bump this sleep to 60 seconds. We don't need to be filling the logs with lots of info messages.

File: tests/cli/test_compose_tar_kickstart.sh [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+chmod +x /etc/rc.d/rc.local
+curl http://10.0.2.2/image-$uuid-installed
+%end
+__EOF__
+        mv $image $image_path
+        restorecon $image_path
+        $QEMU_BIN -m 2048 -drive file=disk.img,format=raw -nographic -kernel vmlinuz -initrd initrd.img \
+            -append "inst.ks=http://10.0.2.2/ks-tar.cfg inst.stage2=$baseurl console=ttyS0" --no-reboot &
+
+        timeout=60 # wait up to 60 minutes
```
Comment (atodorov) [inline tests/cli/test_compose_tar_kickstart.sh]: IMO 60 minutes is too much to wait, maybe something like 30 mins to wait for installation. Eitherway it takes around 20-25 min to build the image and the test will timeout when run in CI before the current timeout is exhausted.

File: tests/cli/test_compose_tar_kickstart.sh [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+        t=0
+        while ((t < timeout)) && ! grep -q "image-$uuid-installed" /var/log/httpd/access_log; do
+            rlLogInfo "Waiting for the tar image to be installed."
+            sleep 1m
+            ((t += 1))
+        done
+        rlRun -t -c "grep image-$uuid-installed /var/log/httpd/access_log" 0 "Installation of the image finished."
```
Comment (atodorov) [inline tests/cli/test_compose_tar_kickstart.sh]: You can also add a rlLogInfo inside the if condition below to avoid grepping twice for the same thing.

File: tests/cli/test_compose_tar_kickstart.sh [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+
+        # rather check explicitly that the installation finished than wait unnecessarily if the installation failed
+        if grep -q "image-$uuid-installed" /var/log/httpd/access_log; then
+            # wait for the installation process to complete and boot the installed system
+            sleep 1m
```
Comment (atodorov) [inline tests/cli/test_compose_tar_kickstart.sh]: Why sleep again ?

File: tests/cli/test_compose_tar_kickstart.sh [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+        while ((t < timeout)) && ! grep -q "image-$uuid-installed" /var/log/httpd/access_log; do
+            rlLogInfo "Waiting for the tar image to be installed."
+            sleep 1m
+            ((t += 1))
+        done
+        rlRun -t -c "grep image-$uuid-installed /var/log/httpd/access_log" 0 "Installation of the image finished."
+
+        # rather check explicitly that the installation finished than wait unnecessarily if the installation failed
+        if grep -q "image-$uuid-installed" /var/log/httpd/access_log; then
+            # wait for the installation process to complete and boot the installed system
```
Comment (atodorov) [inline tests/cli/test_compose_tar_kickstart.sh]: Comment doesn't look correct. We're not waiting for anything here, are we? From what I can see installation mush have finished already.

File: tests/cli/test_compose_tar_kickstart.sh [hunk 15]
```diff
@@ -0,0 +1,158 @@
...
+            sleep 1m
+            $QEMU_BIN -m 2048 -drive file=disk.img,format=raw -nographic -net user,id=nic0,hostfwd=tcp::2222-:22 \
+                -net nic &
+            t=0
+            timeout=10
+            while ((t < timeout)) && ! grep -q "image-$uuid-booted" /var/log/httpd/access_log; do
+                rlLogInfo "Waiting for the installed system to boot."
+                sleep 1m
+                ((t += 1))
+            done
```
Comment (atodorov) [inline tests/cli/test_compose_tar_kickstart.sh]: Cherry-pick this commit
https://github.com/weldr/lorax/pull/819/commits/2a23b36150576fffd8c6d6fbad1ddbb0b48d12d6

and use the boot_image function

File: tests/cli/test_compose_tar_kickstart.sh [hunk 16]
```diff
@@ -0,0 +1,158 @@
...
+            rlRun -t -c "grep image-$uuid-booted /var/log/httpd/access_log" 0 "Installed system booted."
```
Comment (atodorov) [inline tests/cli/test_compose_tar_kickstart.sh]: This will become obsolete with the commit from Brian.

File: test/check-cli [hunk 17]
```diff
@@ -28,6 +28,9 @@ class TestImages(composertest.ComposerTestCase):
     def test_tar(self):
         self.runCliTest("/tests/cli/test_compose_tar.sh")
 
+    def test_tar_kickstart(self):
+        self.runCliTest("/tests/cli/test_compose_tar_kickstart.sh")
+
```
Comment (atodorov) [inline test/check-cli]: Currently the fedora-30 scenario times out so I think tar tests should be separated into their own scenario.

**Comment by atodorov** [comment]
> I saw a comment about IP addresses, and can't tell if it was answered. You can set the subnet to use by passing in net=192.168.103.0/24 as one of the netdev options.

@jikortus what about this comment? where does the `10.0.2.2` IP address come from ?

File: tests/cli/test_compose_tar_kickstart.sh (hunk 11 above)
Comment (atodorov) [reply tests/cli/test_compose_tar_kickstart.sh]: Actually, can you use `cmdline` in kickstart so we can see the progress in the log/terminal and also not start the qemu process in the background. Isn't it possible to block while the installation is running and then shutdown the VM/qemu ? From what I see this is exactly what you are doing (shutting down the VM, then boot it again) so there should be no need to start qemu in the background.

File: tests/cli/test_compose_tar_kickstart.sh (hunk 8 above)
Comment (jikortus) [reply tests/cli/test_compose_tar_kickstart.sh]: OK, I'll fix this.

File: tests/cli/test_compose_tar_kickstart.sh (hunk 9 above)
Comment (jikortus) [reply tests/cli/test_compose_tar_kickstart.sh]: OK, I'll fix this.

File: tests/cli/test_compose_tar_kickstart.sh (hunk 10 above)
Comment (jikortus) [reply tests/cli/test_compose_tar_kickstart.sh]: OK, I'll fix this.

File: tests/cli/test_compose_tar_kickstart.sh (hunk 11 above)
Comment (jikortus) [reply tests/cli/test_compose_tar_kickstart.sh]: OK, I'll lower the timeout (however in some cases that I observed, the installation took quite a long time (~45 minutes, possibly even more), so I hope that 30 minutes will be enough). I'll also change the installation type to cmdline.

However, as for the VM execution/teardown changes, I'm not sure if I understand what you suggest. The thing is that I have to start the machine first with direct kernel boot, perform the installation (and wait for the installation process to wholly finish and shutdown the machine). I can't just reboot, as it's not possible (or I couldn't find a suitable way) to boot from disk after the reboot. Technically I could overcome this by once booting from network (TFTP/iPXE) and then from the local drive, however I think this would be an unnecessary complication.

And if I ran the VM in the foreground, then I'd have to find some other way to manage for the timeout (I guess I'd have to run the timeout check in background, would this approach help us in any way?).

File: tests/cli/test_compose_tar_kickstart.sh (hunk 13 above)
Comment (jikortus) [reply tests/cli/test_compose_tar_kickstart.sh]: This is meant to give the installation some time to finish correctly (and in turn the qemu-kvm process as well), before we can safely start the second qemu-kvm process (with boot from local drive).

File: tests/cli/test_compose_tar_kickstart.sh (hunk 12 above)
Comment (jikortus) [reply tests/cli/test_compose_tar_kickstart.sh]: Good idea, thanks.

File: tests/cli/test_compose_tar_kickstart.sh (hunk 14 above)
Comment (jikortus) [reply tests/cli/test_compose_tar_kickstart.sh]: Yes, I could use a better wording here. The installation process is finished at this moment (everything is done, including the %post scripts), but we have to wait for a while for the qemu-kvm process to finish after the installation is complete (as I commented above for the sleep).

File: tests/cli/test_compose_tar_kickstart.sh (hunk 11 above)
Comment (jikortus) [reply tests/cli/test_compose_tar_kickstart.sh]: Based on our yesterday's conversation, I performed the required changes for the qemu-kvm process to be blocking (and the timeout of the test will be managed by CI).

File: tests/cli/test_compose_tar_kickstart.sh (hunk 12 above)
Comment (jikortus) [reply tests/cli/test_compose_tar_kickstart.sh]: I believe this is no more relevant, WRT changes made to how the VM is run.

File: tests/cli/test_compose_tar_kickstart.sh (hunk 13 above)
Comment (jikortus) [reply tests/cli/test_compose_tar_kickstart.sh]: No more relevant as well.

File: tests/cli/test_compose_tar_kickstart.sh (hunk 14 above)
Comment (jikortus) [reply tests/cli/test_compose_tar_kickstart.sh]: No more relevant as well.

File: tests/cli/test_compose_tar_kickstart.sh (hunk 15 above)
Comment (jikortus) [reply tests/cli/test_compose_tar_kickstart.sh]: Done.

File: test/check-cli (hunk 17 above)
Comment (jikortus) [reply test/check-cli]: Done.

File: tests/cli/test_compose_tar_kickstart.sh (hunk 16 above)
Comment (jikortus) [reply tests/cli/test_compose_tar_kickstart.sh]: > > I saw a comment about IP addresses, and can't tell if it was answered. You can set the subnet to use by passing in net=192.168.103.0/24 as one of the netdev options.
> 
> @jikortus what about this comment? where does the `10.0.2.2` IP address come from ?

Sorry that I missed this comment. 10.0.2.2 is the default host IP address used by the internal qemu-kvm DHCP server (and FTR, 10.0.2.15 is the first guest address in the range), as I found out in the documentation. Of course this can be changed to any custom address range, but I think the defaults should work well for us.

Comment (atodorov) [reply tests/cli/test_compose_tar_kickstart.sh]: OK, if the docs say this is the default it's fine.

File: tests/cli/test_compose_tar_kickstart.sh [hunk 18]
```diff
@@ -0,0 +1,133 @@
...
+
+__EOF__
+        mv $image $image_path
+        restorecon $image_path
+        rlLogInfo "Starting installation from tar image in a VM"
+        $QEMU_BIN -m 2048 -drive file=disk.img,format=raw -nographic -kernel vmlinuz -initrd initrd.img \
+            -append "inst.ks=http://10.0.2.2/ks-tar.cfg inst.stage2=$baseurl console=ttyS0" --no-reboot
+
+        rlLogInfo "Installation of the image finished."
+        boot_image "-drive file=disk.img,format=raw" 600
```
Comment (atodorov) [inline tests/cli/test_compose_tar_kickstart.sh]: You can split this into a separate rlPhase to visually distinguish between installation and boot+verify of the VM

File: tests/cli/test_compose_live-iso.sh [hunk 19]
```diff
@@ -13,8 +13,6 @@ set -e
 . $(dirname $0)/lib/lib.sh
 
 CLI="${CLI:-./src/bin/composer-cli}"
-QEMU_BIN="/usr/bin/qemu-system-$(uname -m)"
-QEMU="$QEMU_BIN -machine accel=kvm:tcg"
```
[the remark this answers is not in the document]
Comment (jikortus) [reply tests/cli/test_compose_live-iso.sh]: OK, I fixed that. However, is this still needed [in general], as we moved to using the separate function for starting the VM? Btw, now that I looked again at the cherry-picked commit, is `export QEMU="/usr/libexec/qemu-kvm"` what we really want on Fedora? Unforntunately I haven't been able yet to go past the installation step in this test on Fedora (bugs during installation on F-30, issues with packages downloading on rawhide), so that's probably why I haven't noticed this sooner.

File: tests/cli/test_compose_tar_kickstart.sh (hunk 18 above)
Comment (jikortus) [reply tests/cli/test_compose_tar_kickstart.sh]: OK, good idea.

File: tests/cli/test_compose_live-iso.sh (hunk 19 above)
[the remark this answers is not in the document]
Comment (atodorov) [reply tests/cli/test_compose_live-iso.sh]: I don't see this to have changed. On Fedora test systems there's `qemu-system-x86_64` but no `qemu-kvm` so the qcow2 test fails.

File: test/check-cli [hunk 20]
```diff
@@ -39,5 +36,13 @@ class TestLiveIso(composertest.ComposerTestCase):
         self.runCliTest("/tests/cli/test_compose_live-iso.sh")
 
 
+class TestTarKickstart(composertest.ComposerTestCase):
```
[the remark this answers is not in the document]
Comment (jikortus) [reply test/check-cli]: OK, I missed this one.

**Comment by jikortus** [comment]
Commits squashed. As for the commented `QEMU*` variable lines in `test_compose_live-iso.sh` - I'm not sure what exactly you meant in your comment.

My guess is that the best option would be to remove those lines, use the new `boot_image` function to boot the iso image, and change `$QEMU_BIN` in `tests/cli/lib/lib.sh`  to `qemu-system-$(uname -m)`, just for the master branch (and leave `/usr/libexec/qemu-kvm` for rhel-7 and 8). Is this correct?

File: tests/cli/lib/lib.sh [hunk 21]
```diff
@@ -16,6 +16,27 @@ if [ "$COMPOSER_TEST_FAIL_FAST" == "1" ]; then
   }
 fi
 
+export QEMU="qemu-system-$(uname -m)"
```
Comment (atodorov) [inline tests/cli/lib/lib.sh]: This is missing the additional qemu options from the test scripts before the cherry-pick. Just copy the deleted lines and paste them here.

File: tests/cli/test_compose_qcow2.sh [hunk 22]
```diff
@@ -13,12 +13,10 @@ set -e
 . $(dirname $0)/lib/lib.sh
 
 CLI="${CLI:-./src/bin/composer-cli}"
-QEMU_BIN="/usr/bin/qemu-system-$(uname -m)"
-QEMU="$QEMU_BIN -machine accel=kvm:tcg"
 
 rlJournalStart
     rlPhaseStartSetup
-        rlAssertExists $QEMU_BIN
+        rlAssertExists $(which $QEMU)
```
Comment (atodorov) [inline tests/cli/test_compose_qcow2.sh]: If you copy the deleted QEMU and QEMU_BIN variables and adjust their values in the cherry-picked commit all of these assert changes will not be needed.

File: tests/cli/test_compose_tar_kickstart.sh [hunk 23]
```diff
@@ -0,0 +1,133 @@
...
+#####
+
+set -e
+
+. /usr/share/beakerlib/beakerlib.sh
+. $(dirname $0)/lib/lib.sh
+
+CLI="${CLI:-./src/bin/composer-cli}"
+QEMU="qemu-system-$(uname -m)"
+QEMU_CMD="$QEMU -machine accel=kvm:tcg"
```
Comment (atodorov) [inline tests/cli/test_compose_tar_kickstart.sh]: These should not be needed, right? Why not use QEMU which is defined in the shared library ?

File: tests/cli/test_compose_tar_kickstart.sh [hunk 24]
```diff
@@ -0,0 +1,133 @@
...
+description = "tar image test"
+version = "0.0.1"
+modules = []
+
+[[packages]]
+name = "openssh-server"
+version = "*"
+
+[[groups]]
+name = "anaconda-tools"
```
Comment (atodorov) [inline tests/cli/test_compose_tar_kickstart.sh]: this shouldn't be needed with the new image type

File: tests/cli/lib/lib.sh (hunk 21 above)
Comment (jikortus) [reply tests/cli/lib/lib.sh]: OK, I changed the QEMU* variables to conform to the previous state (before Brian's commit), hope it's fine this way.

File: tests/cli/test_compose_tar_kickstart.sh (hunk 23 above)
Comment (jikortus) [reply tests/cli/test_compose_tar_kickstart.sh]: Not exactly in accordance to the variables from Brian's commit, but I changed the variables to $QEMU_BIN (path to the qemu-system binary) and $QEMU (qemu-system binary with the accelerator parameters), and therefore there is no need to define them anymore in the test script.

File: tests/cli/test_compose_tar_kickstart.sh (hunk 24 above)
Comment (jikortus) [reply tests/cli/test_compose_tar_kickstart.sh]: I wasn't 100% sure about this one. But if the reason `@anaconda-tools` group needed to be included was `grub2` and `grub2-tools`, then it can be safely removed from the blueprint.

Comment (atodorov) [reply tests/cli/test_compose_tar_kickstart.sh]: well Brian says the new image type will automatically add everything needed so we don't need to add anything extra on our own. Remove the groups section, tests should continue working.

**Comment by jikortus** [comment]
@atodorov OK, then it's fine to remove the `groups` section. As for the `partitioned_disk` test, we'd need the lorax-composer logs to investigate why it failed...

**Comment by atodorov** [comment]
@bcl with the new image type & test we're getting the following error:

> No such file or directory: 'new-kernel-pkg'

Looks like we also need `grubby` in the package list, no ?

**Comment by bcl** [comment]
> Looks like we also need `grubby` in the package list, no ?

No, that should be fixed by the latest anaconda with the liveimg patches. When using reqpart in the template and the newest anaconda there is no need to add extra packages to handle booting, anaconda will do that automatically.

**Comment by atodorov** [comment]
> No, that should be fixed by the latest anaconda with the liveimg patches. When using reqpart in the template and the newest anaconda there is no need to add extra packages to handle booting, anaconda will do that automatically.

so we just need to wait for a newer anaconda then. Any ETA on that ?

**Comment by bcl** [comment]
> so we just need to wait for a newer anaconda then. Any ETA on that ?

Looks like it should already be available in 31.19 and 30.25.6-4

**Comment by atodorov** [comment]
Indeed the updates repo has 30.25.6-4 but the installation process is using anaconda 30.25.6-2.fc30, probably because initrd image has not been refreshed. 

I don't know if this can be refreshed or if we need to switch to Fedora-31 images/repositories for testing (if those are available).

Otherwise this will continue to fail so we've got to skip it until a newer anaconda is available in stage2.

**Comment by atodorov** [comment]
@bcl this is ready for merge. Few notes:

1) `fedora-30` failure is caused by commit 3b6c6a3 which has been reverted in #836
2) The actual kickstart tar test has been skipped b/c Fedora 30 repositories have an older anaconda in stage2 which doesn't seem likely to be refreshed
3) `continuous-integration/travis-ci/pr` failure has nothing to do with the current changes

B/c there are failing CI statuses which are marked as Required I can't merge myself. If you don't have any other comments please merge.

**Comment by atodorov** [comment]
@bcl, @larskarlitski ping. Is there anything else to do here ? Please review/merge.

File: tests/cli/test_compose_tar_kickstart.sh [hunk 25]
```diff
@@ -0,0 +1,128 @@
...
+# reboot is used together with --no-reboot qemu-kvm parameter, which makes the qemu-kvm
+# process exit after the installation is complete and anaconda reboots the system
+# (using 'poweroff' ks command just halted the machine without powering it off)
+reboot
+
+liveimg --url http://10.0.2.2/root.tar.xz
+
+__EOF__
+        mv $image $image_path
+        restorecon $image_path
```
Comment (larskarlitski) [inline tests/cli/test_compose_tar_kickstart.sh]: Why is this necessary? Please add a comment.

File: tests/cli/test_compose_tar_kickstart.sh [hunk 26]
```diff
@@ -0,0 +1,128 @@
...
+        rlLogInfo "Starting installation from tar image in a VM"
+        $QEMU -m 2048 -drive file=disk.img,format=raw -nographic -kernel vmlinuz -initrd initrd.img \
+            -append "inst.ks=http://10.0.2.2/ks-tar.cfg inst.stage2=$baseurl console=ttyS0" --no-reboot
+
+        rlLogInfo "Installation of the image finished."
```
Comment (larskarlitski) [inline tests/cli/test_compose_tar_kickstart.sh]: How do we check that the installation succeeded?

File: tests/cli/test_compose_tar_kickstart.sh [hunk 27]
```diff
@@ -0,0 +1,128 @@
...
+    rlPhaseEnd
+
+    rlPhaseStartTest "Boot and check the installed system"
+        boot_image "-drive file=disk.img,format=raw" 600
+        # run generic tests to verify the instance
+        CHECK_CMDLINE=0 verify_image root localhost "-i $ssh_key_dir/id_rsa -p 2222"
+    rlPhaseEnd
+
+    rlPhaseStartCleanup
+        rlRun -t -c "killall -9 $QEMU_BIN"
```
Comment (larskarlitski) [inline tests/cli/test_compose_tar_kickstart.sh]: Why is this necessary? qemu has exited in the previous phase, no?

File: tests/cli/test_compose_tar_kickstart.sh (hunk 26 above)
Comment (atodorov) [reply tests/cli/test_compose_tar_kickstart.sh]: if qemu doesn't exit here there's probably an error which happened. 

In most cases when we get here installation would have completed successfully, although not 100% guaranteed. The way we actually check for this is a bit later when we try to boot from the freshly installed image and verify some artifacts inside the running VM.

AFAIK there's no easy way to check that the installation process trully completed without any error. Some of the installation tests by RTT collect data during the install and later verify this data when the system boots. If something goes wrong they usually time out.

File: tests/cli/test_compose_tar_kickstart.sh (hunk 25 above)
Comment (jikortus) [reply tests/cli/test_compose_tar_kickstart.sh]: The built image doesn't have a proper security context for it to be accessed and served by Apache, so the context has to be restored to a proper one in `/var/www/html`.

**Comment by atodorov** [comment]
Rebased for the failure in `fedora-30` scenario. Fixed in 15818421327dfb46d6d74a65f1a617a641fa5431

File: tests/cli/test_compose_tar_kickstart.sh (hunk 26 above)
Comment (jikortus) [reply tests/cli/test_compose_tar_kickstart.sh]: Yes, it works exactly how Alex described it. I think this approach is sufficient enough, as any issue during installation (traceback, incomplete spoke, some other error) would most probably lead to a timeout of the test, otherwise it would be quite difficult to check if installation goes on well. It would be rather more useful to check the system after installation, if necessary.

File: tests/cli/test_compose_tar_kickstart.sh (hunk 27 above)
Comment (jikortus) [reply tests/cli/test_compose_tar_kickstart.sh]: No, because `boot_image` runs qemu as a background process, so that it can be checked afterwards. Therefore we need to kill it in the cleanup phase as it's still running.

**Comment by jikortus** [comment]
@larskarlitski Thank you for the review! I rebased to master, but ATM several tests have failed on me, looks like an issue with the test VM image (image-download).

**Comment by jikortus** [comment]
@larskarlitski All of the tests are green after I restarted them, could you please merge this PR if you don't have any further objections or suggestions? :)
